### PR TITLE
fix(iam): incorrect input for AccountPrincipal is building successfully 

### PR DESCRIPTION
**Pull request by duranbe**
### Reason for this change

There is no validation and test that the AWS Account Id when creating AccountPrincipal Object. In my case I missed a digit when copy pasting an account id and the build still passed, the typo has been caught only during the code review process

### Description of changes

Adding simple regex to check that AWS Id is 12 digits long & update error message

### Description of how you validated changes

Existing tests are passing and added 2 of them

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by duranbe** [comment]
Exemption Request

**Comment by duranbe** [comment]
PR small enough - no need to update integ test (IMHO)

**Comment by duranbe** [comment]
Logs are quite odd 

```
[4/4] Building fresh packages...
error /codebuild/output/src3225815236/src/github.com/aws/aws-cdk/node_modules/@lerna/create/node_modules/nx, /codebuild/output/src3225815236/src/github.com/aws/aws-cdk/node_modules/@nx/devkit/node_modules/nx, /codebuild/output/src3225815236/src/github.com/aws/aws-cdk/node_modules/lerna/node_modules/nx: Command failed.
Exit code: 135
Command: node ./bin/post-install
Arguments: 
Directory: /codebuild/output/src3225815236/src/github.com/aws/aws-cdk/node_modules/@lerna/create/node_modules/nx
Output:
Bus error (core dumped)
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.

```
Seems issue with yarn/nx

File: packages/aws-cdk-lib/aws-iam/lib/principals.ts [hunk 1]
```diff
@@ -473,14 +473,16 @@ export class ArnPrincipal extends PrincipalBase {
...
+
   /**
    *
    * @param accountId AWS account ID (i.e. '123456789012')
    */
   constructor(public readonly accountId: any) {
     super(new StackDependentToken(stack => `arn:${stack.partition}:iam::${accountId}:root`).toString());
-    if (!cdk.Token.isUnresolved(accountId) && typeof accountId !== 'string') {
-      throw new Error('accountId should be of type string');
+    if (!cdk.Token.isUnresolved(accountId) && typeof accountId !== 'string' && !this.accountIdRegExp.test(accountId)) {
```
Comment (TheRealAmazonKendra) [inline packages/aws-cdk-lib/aws-iam/lib/principals.ts:484]: While you're at it, maybe we should check to see that all characters in the string are digits.

**Review by TheRealAmazonKendra — DISMISSED** [review]

**Comment by TheRealAmazonKendra** [comment]
This build failure can't possibly have anything to do with this change... I'll retrigger it.

**Comment by TheRealAmazonKendra** [comment]
@Mergifyio update

**Comment by TheRealAmazonKendra** [comment]
For fixes, the title should describe the bug, not the solution. Could you please update it?

File: packages/aws-cdk-lib/aws-iam/lib/principals.ts (hunk 1 above)
Comment (duranbe) [reply packages/aws-cdk-lib/aws-iam/lib/principals.ts:484]: It's covered with the regex `new RegExp('^[0-9]{12}$');`

**Comment by duranbe** [comment]
Hey Kendra, thanks for the review and restarting the build 

> For fixes, the title should describe the bug, not the solution. Could you please update it?

Done ! Is it better with this ?

**Comment by duranbe** [comment]
Seems this PR is breaking lots of existing test ... I'll look into it, if it's too complex or there is actual use case where the an AWS ID is not 12 digits (which should not be the case ?) I'll close it.

File: packages/@aws-cdk-testing/framework-integ/test/aws-codepipeline-actions/test/integ.pipeline-ecr-source.ts [hunk 2]
```diff
@@ -15,9 +15,9 @@ import * as cpactions from 'aws-cdk-lib/aws-codepipeline-actions';
...
- * aws-cdk-codepipeline-ecr-source.LoginCommand = aws ecr get-login-password --region us-east-1 | docker login --username AWS --password-stdin 11111111111.dkr.ecr.us-east-1.amazonaws.com
- * aws-cdk-codepipeline-ecr-source.PushCommand = docker tag public.ecr.aws/lambda/provided 11111111111.dkr.ecr.us-east-1.amazonaws.com/aws-cdk-codepipeline-ecr-source-myecrrepo767466d0-gsrntpvfwc5w:latest \
- * && docker push 11111111111.dkr.ecr.us-east-1.amazonaws.com/aws-cdk-codepipeline-ecr-source-myecrrepo767466d0-gsrntpvfwc5w:latest
+ * aws-cdk-codepipeline-ecr-source.LoginCommand = aws ecr get-login-password --region us-east-1 | docker login --username AWS --password-stdin 111111111111.dkr.ecr.us-east-1.amazonaws.com
+ * aws-cdk-codepipeline-ecr-source.PushCommand = docker tag public.ecr.aws/lambda/provided 111111111111.dkr.ecr.us-east-1.amazonaws.com/aws-cdk-codepipeline-ecr-source-myecrrepo767466d0-gsrntpvfwc5w:latest \
+ * && docker push 111111111111.dkr.ecr.us-east-1.amazonaws.com/aws-cdk-codepipeline-ecr-source-myecrrepo767466d0-gsrntpvfwc5w:latest
```
Comment (duranbe) [inline packages/@aws-cdk-testing/framework-integ/test/aws-codepipeline-actions/test/integ.pipeline-ecr-source.ts:20]: might not be needed tho

File: packages/@aws-cdk-testing/framework-integ/test/aws-codepipeline-actions/test/cloudformation/integ.stacksets.ts [hunk 3]
```diff
@@ -53,7 +53,7 @@ export class StackSetPipelineStack extends Stack {
       ],
     });
 
-    const accounts = process.env.STACKSET_ACCOUNTS?.split(',') ?? ['1111', '2222'];
+    const accounts = process.env.STACKSET_ACCOUNTS?.split(',') ?? ['1111', '222222222222'];
```
Comment (duranbe) [inline packages/@aws-cdk-testing/framework-integ/test/aws-codepipeline-actions/test/cloudformation/integ.stacksets.ts:56]: need to check if needed

**Comment by duranbe** [comment]
Had to refactor/fix in bunch of test files but now they use proper mock AWS Accound id 🙌

**Comment by duranbe** [comment]
Bump !

**Comment by duranbe** [comment]
Good on my side :) 👍

**Comment by duranbe** [comment]
@TheRealAmazonKendra can you take a quick look ? PR seems big (🥲) but 99% is just updating existing test

**Pull request closed without merging**